thinc/loss.py

```diff
@@ -143,6 +148,25 @@ def get_grad(self, guesses: Floats2d, truths: IntsOrFloatsOrStrs) -> Floats2d:
             raise ValueError(err)
         difference = guesses - target
         difference *= mask
+        # Weight samples proportinal to the marginals.
+        # For each sample the target labels are normalized
+        # to sum to 1 and the normalized-labels are used
+        # to compute the weighted average of class_weights
+        if self.class_weights is not None:
+            if target.shape[1] != len(self.class_weights):
+                raise ValueError(
+                    "The number of classes in the "
+                    "target and in class weights "
+                    "has to be equal, but found "
+                    f"{target.shape[1]} and "
+                    f"{len(self.class_weights)}"
+                )
+            cp = xp.asarray(self.class_weights)
+            cp = xp.tile(cp, (guesses.shape[0], 1))
+            row_sum = xp.expand_dims(xp.sum(target, axis=1), axis=1)
+            norm_target = target / row_sum
+            sample_weights = (norm_target * cp).sum(axis=1)
+            difference *= xp.expand_dims(sample_weights, 1)
```

Comment:
This `expand_dims` may also not be necessary by using the `keepdims` argument of `sum`.